src/Adapter/MSTest.CoreAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -220,11 +196,191 @@ internal string GetLoadExceptionDetails(ReflectionTypeLoadException ex)
         /// <returns>a TypeEnumerator instance.</returns>
         internal virtual TypeEnumerator GetTypeEnumerator(Type type, string assemblyFileName)
         {
-            var reflectHelper = new ReflectHelper();
-            var typevalidator = new TypeValidator(reflectHelper);
-            var testMethodValidator = new TestMethodValidator(reflectHelper);
+            var typeValidator = new TypeValidator(ReflectHelper);
+            var testMethodValidator = new TestMethodValidator(ReflectHelper);
+
+            return new TypeEnumerator(type, assemblyFileName, ReflectHelper, typeValidator, testMethodValidator);
+        }
+
+        private IEnumerable<UnitTestElement> DiscoverTestsInType(string assemblyFileName, string runSettingsXml, Assembly assembly, Type type, List<string> warningMessages)
+        {
+            var sourceLevelParameters = PlatformServiceProvider.Instance.SettingsProvider.GetProperties(assemblyFileName);
+            sourceLevelParameters = RunSettingsUtilities.GetTestRunParameters(runSettingsXml)?.ConcatWithOverwrites(sourceLevelParameters)
+                ?? sourceLevelParameters
+                ?? new Dictionary<string, object>();
+
+            string typeFullName = null;
+            var tests = new List<UnitTestElement>();
+
+            try
+            {
+                typeFullName = type.FullName;
+                var unitTestCases = this.GetTypeEnumerator(type, assemblyFileName).Enumerate(out var warningsFromTypeEnumerator);
+                var typeIgnored = ReflectHelper.IsAttributeDefined(type, typeof(UTF.IgnoreAttribute), false);
+
+                if (warningsFromTypeEnumerator != null)
+                {
+                    warningMessages.AddRange(warningsFromTypeEnumerator);
+                }
+
+                if (unitTestCases != null)
+                {
+                    foreach (var test in unitTestCases)
+                    {
+                        test.Ignored = typeIgnored || test.Ignored;
+                        if (test.Ignored)
+                        {
+                            tests.Add(test);
+                            continue;
+                        }
+
+                        if (this.DynamicDataAttached(sourceLevelParameters, assembly, test, tests))
+                        {
+                            continue;
+                        }
+
+                        tests.Add(test);
```

Comment:
So, should we discover all tests, even if they are ignored or in an ignored class?

Reply:
I think we should discover all tests. Ignored tests should be discovered, but should return the "skipped" status upon execution (and should be displayed with a yellow skipped icon in VS). This would match the behavior for other frameworks.

If we skip discovery for ignored tests they would not be displayed in VS. This means it would be impossible to un-skip and execute them again without running a build to make them appear first after un-skipping.

One may also want to know how many tests in a given solution are currently ignored by looking at total count of ignored tests in VS (and fix them up / un-skip by navigating to them one-by-one). This would again not be impossible if the skipped tests are not discovered in the first place.

Some frameworks also have a conditional skip attribute (e.g. skip if OS=Linux) where the condition is evaluated at run time. I am not sure if mstest has such capability - but this would be another reason to only skip at run time.